src/prefect/_internal/concurrency/workers.py

```diff
@@ -0,0 +1,227 @@
+        # Do not execute if the future is cancelled
+        if not self.future.set_running_or_notify_cancel():
+            logger.debug("Skipping execution of cancelled call %r", self)
+            return
+
+        logger.debug("Running call %r", self)
+
+        coro = self._run_sync()
+        if coro is not None:
+            loop = get_running_loop()
+            if loop:
+                # If an event loop is available, return a task to be awaited
+                # Note we must create a task for context variables to propagate
+                logger.debug(
+                    "Executing coroutine for call %r in running loop %r", self, loop
+                )
+                return self.context.run(loop.create_task, self._run_async(coro))
+            else:
+                # Otherwise, execute the function here
+                logger.debug("Executing coroutine for call %r in new loop", self)
+                return self.context.run(asyncio.run, self._run_async(coro))
+
+        return None
+
+    def _run_sync(self):
+        try:
+            result = self.context.run(self.fn, *self.args, **self.kwargs)
+
+            # Return the coroutine for async execution
+            if inspect.isawaitable(result):
+                return result
+
+        except BaseException as exc:
+            self.future.set_exception(exc)
+            logger.debug("Encountered exception in call %r", self)
+            # Prevent reference cycle in `exc`
+            del self
+        else:
+            self.future.set_result(result)
+            logger.debug("Finished call %r", self)
+
+    async def _run_async(self, coro):
+        try:
+            result = await coro
+        except BaseException as exc:
+            logger.debug("Encountered exception %s in async call %r", exc, self)
+            self.future.set_exception(exc)
+            # Prevent reference cycle in `exc`
+            del self
+        else:
+            self.future.set_result(result)
+            logger.debug("Finished async call %r", self)
+
+    def __call__(self) -> T:
+        """
+        Execute the call and return its result.
+
+        All executions during excecution of the call are re-raised.
+        """
+        coro = self.run()
+
+        # Return an awaitable if in an async context
+        if coro is not None:
+
+            async def run_and_return_result():
+                await coro
+                return self.future.result()
+
+            return run_and_return_result()
+        else:
+            return self.future.result()
+
+    def __repr__(self) -> str:
+        name = getattr(self.fn, "__name__", str(self.fn))
+        call_args = ", ".join(
+            [repr(arg) for arg in self.args]
+            + [f"{key}={repr(val)}" for key, val in self.kwargs.items()]
+        )
+
+        # Enforce a maximum length
+        if len(call_args) > 100:
+            call_args = call_args[:100] + "..."
+
+        return f"{name}({call_args})"
+
+
+class Worker:
```

Comment:
Are we asking for trouble naming this `Worker`?  I know it's private, but searching the code for `Worker` is going to be sampling from two disjoint worlds (`Worker` the product feature and `Worker` the concurrency implementation detail).  Perhaps `Minion`? :D

Reply:
I was thinking about that too. I'll try to think of a name that doesn't overlap.